src/libutil/file-path-impl.hh

```diff
@@ -0,0 +1,74 @@
+#pragma once
+/**
+ * @file
+ *
+ * Pure (no IO) infrastructure just for defining other path types;
+ * should not be used directly outside of utilities.
+ */
+#include <string>
+#include <string_view>
+
+namespace nix {
+
+/**
+ * Core pure path canonicalization algorithm.
+ *
+ * @param hookComponent A chance to modify `s` and `path` in an
+ * arbitrary way, e.g. if it points to a symlink. Does nothing by
+ * default.
+ */
+typename std::string canonPathInner(
+    std::string_view path,
+    auto && hookComponent)
```

Comment:
Can't this be made a `std::function<void(std::string &, std::string_view &)>` to avoid the need for a template?

Reply:
Actually, I'd rather make this a flag to follow or not symlinks (unless there's a reason to generalize it). Higher-order functions are nice but add some non-trivial complexity (a consumer would have a pretty hard time re-implementing the `followSymlinks` lambda used here) so I don't think it's a great tradeoff here

Reply:
To quote the PR description

> The core `CanonPath` constructors were using `absPath`, but `absPath` in some situations does IO which is not appropriate. It turns out that these constructors avoided those situations, and thus were pure, but it was far from obvious this was the case.

Reply:
Also fwiw in #9767 this code needs template regardless; I didn't just do it to try to make the function inline.

Reply:
My interpretation of https://github.com/NixOS/nix/pull/9767#issuecomment-1927217878 and https://discourse.nixos.org/t/2024-02-05-nix-team-meeting-minutes-121/39340 is that we do want to do that PR, and so this the stepping-stone.